tests-integration/tests/integration/smt/matrix/flatten-no-depth/input-expected-rule-trace-human.txt

```diff
@@ -0,0 +1,280 @@
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = b[4]) 
+
+--
+
+flatten(a)[2], 
+   ~~> index_to_bubble ([("Bubble", 6000)]) 
+{flatten(a)[2] @ and([__inDomain(2,int(1..4));int(1..)])} 
+
+--
+
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+({flatten(a)[2] @ and([__inDomain(2,int(1..4));int(1..)])} = b[2]),
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = b[4]), 
+   ~~> constant_evaluator ([("Constant", 9001)]) 
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = b[4]) 
+
+--
+
+b[2], 
+   ~~> index_to_bubble ([("Bubble", 6000)]) 
+{b[2] @ and([__inDomain(2,int(1..4));int(1..)])} 
+
+--
+
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = {b[2] @ and([__inDomain(2,int(1..4));int(1..)])}),
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = b[4]), 
+   ~~> constant_evaluator ([("Constant", 9001)]) 
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = b[4]) 
+
+--
+
+flatten(a)[3], 
+   ~~> index_to_bubble ([("Bubble", 6000)]) 
+{flatten(a)[3] @ and([__inDomain(3,int(1..4));int(1..)])} 
+
+--
+
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+({flatten(a)[3] @ and([__inDomain(3,int(1..4));int(1..)])} = b[3]),
+(flatten(a)[4] = b[4]), 
+   ~~> constant_evaluator ([("Constant", 9001)]) 
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = b[4]) 
+
+--
+
+b[3], 
+   ~~> index_to_bubble ([("Bubble", 6000)]) 
+{b[3] @ and([__inDomain(3,int(1..4));int(1..)])} 
+
+--
+
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+(flatten(a)[3] = {b[3] @ and([__inDomain(3,int(1..4));int(1..)])}),
+(flatten(a)[4] = b[4]), 
+   ~~> constant_evaluator ([("Constant", 9001)]) 
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = b[4]) 
+
+--
+
+flatten(a)[4], 
+   ~~> index_to_bubble ([("Bubble", 6000)]) 
+{flatten(a)[4] @ and([__inDomain(4,int(1..4));int(1..)])} 
+
+--
+
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+(flatten(a)[3] = b[3]),
+({flatten(a)[4] @ and([__inDomain(4,int(1..4));int(1..)])} = b[4]), 
+   ~~> constant_evaluator ([("Constant", 9001)]) 
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = b[4]) 
+
+--
+
+b[4], 
+   ~~> index_to_bubble ([("Bubble", 6000)]) 
+{b[4] @ and([__inDomain(4,int(1..4));int(1..)])} 
+
+--
+
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = {b[4] @ and([__inDomain(4,int(1..4));int(1..)])}), 
+   ~~> constant_evaluator ([("Constant", 9001)]) 
+allDiff(flatten(a)),
+(flatten(a)[1] = b[1]),
+(flatten(a)[2] = b[2]),
+(flatten(a)[3] = b[3]),
+(flatten(a)[4] = b[4]) 
+
+--
+
+flatten(a), 
+   ~~> unwrap_flatten_matrix_nonatomic ([("Smt", 999)]) 
+[a[1, 1],a[1, 2],a[2, 1],a[2, 2];int(1..4)] 
```

Comment:
This rule unwraps the whole matrix even though only one element of it is going to be used when indexing into it.

Perhaps add a rule for cases involving flatten within an index expression that does the maths to choose the necessary indices and avoids doing the full enumerate_indices